File: runner/autotune.py
#!/usr/bin/env python3
"""Auto-improvement agent: analyzes runtime health and emits recommendations."""

from __future__ import annotations

import argparse
from datetime import datetime, timezone
import json
import os
from pathlib import Path
import time
from typing import Any

from runner.queue import QueueError, RedisQueue
from shared.policy import PolicyError, load_policy

DEFAULT_CONFIG_DIR = Path("/data/clawbrain/config")
DEFAULT_LOGS_DIR = Path("/data/clawbrain/logs")
DEFAULT_ARTIFACTS_DIR = Path("/data/clawbrain/artifacts")
DEFAULT_REDIS_URL = "redis://127.0.0.1:6379/0"
DEFAULT_QUEUE_NAME = "clawbrain:tasks"
DEFAULT_STATUS_REDIS_KEY = "clawbrain:agent_status:latest"


def utc_now() -> datetime:
    return datetime.now(timezone.utc)


def utc_now_iso() -> str:
    return utc_now().isoformat()


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="ClawBrain AutoTuneAgent")
    parser.add_argument(
        "--config-dir",
        default=os.environ.get("CLAWBRAIN_CONFIG_DIR", str(DEFAULT_CONFIG_DIR)),
    )
    parser.add_argument(
        "--logs-dir",
        default=os.environ.get("CLAWBRAIN_LOGS_DIR", str(DEFAULT_LOGS_DIR)),
    )
    parser.add_argument(
        "--artifacts-dir",
        default=os.environ.get("CLAWBRAIN_ARTIFACTS_DIR", str(DEFAULT_ARTIFACTS_DIR)),
    )
    parser.add_argument(
        "--redis-url",
        default=os.environ.get("CLAWBRAIN_REDIS_URL", DEFAULT_REDIS_URL),
    )
    parser.add_argument(
        "--queue-name",
        default=os.environ.get("CLAWBRAIN_QUEUE_NAME", DEFAULT_QUEUE_NAME),
    )
    parser.add_argument(
        "--status-redis-key",
        default=os.environ.get("CLAWBRAIN_AGENT_STATUS_REDIS_KEY", DEFAULT_STATUS_REDIS_KEY),
    )
    parser.add_argument(
        "--interval-seconds",
        type=int,
        default=int(os.environ.get("CLAWBRAIN_AUTOTUNE_INTERVAL_SEC", "180")),
    )
    parser.add_argument(
        "--max-log-files",
        type=int,
        default=int(os.environ.get("CLAWBRAIN_AUTOTUNE_MAX_LOG_FILES", "40")),
    )
    parser.add_argument("--once", action="store_true")
    return parser.parse_args()


def _summarize_recent_logs(logs_dir: Path, max_files: int) -> dict[str, Any]:
    if not logs_dir.is_dir():
        return {"log_files_count": 0, "total_log_size_bytes": 0, "newest_logs": []}

    entries = sorted(
        [path for path in logs_dir.glob("*.log") if path.is_file()],
        key=lambda path: path.stat().st_mtime,
        reverse=True,
    )
    newest = entries[: max(1, max_files)]

    files_info: list[dict[str, Any]] = []
    total = 0
    for path in newest:
        stat = path.stat()
        total += stat.st_size
        files_info.append(
            {
                "name": path.name,
                "size_bytes": stat.st_size,
                "mtime": datetime.fromtimestamp(stat.st_mtime, tz=timezone.utc).isoformat(),
            }
        )

    return {
        "log_files_count": len(entries),
        "total_log_size_bytes": total,
        "newest_logs": files_info[:10],
    }


def _load_status_snapshot(queue: RedisQueue, status_redis_key: str) -> dict[str, Any]:
    try:
        raw = queue.client.get(status_redis_key)
    except Exception:
        return {}
    if not raw:
        return {}
    try:
        payload = json.loads(raw)
    except json.JSONDecodeError:
        return {}
    return payload if isinstance(payload, dict) else {}


def _build_recommendations(
    *,
    policy: Any | None,
    queue_depth: int,
    status_snapshot: dict[str, Any],
    logs_summary: dict[str, Any],
) -> list[str]:
    recommendations: list[str] = []
    missing_agents = status_snapshot.get("missing_agents") if isinstance(status_snapshot.get("missing_agents"), list) else []

    if queue_depth >= 5:
        recommendations.append("Queue depth >= 5: activar profile reinforce para escalar runner-coder-r2 y runner-builder-r2.")
    elif queue_depth > 0:
        recommendations.append("Hay tareas en cola: monitorear tiempos y mantener AutoTuneAgent activo.")
    else:
        recommendations.append("Cola en cero: estado estable.")

    if missing_agents:
        recommendations.append(f"Agentes faltantes reportados por maintainer: {', '.join(str(x) for x in missing_agents)}.")
    else:
        recommendations.append("Maintainer reporta agentes requeridos en linea.")

    if policy is None:
        recommendations.append("No se pudo cargar policy activa, validar /data/clawbrain/config/policy.yaml.")
    else:
        repos_allowed = list(getattr(policy, "repos_allowed", []))
        stacks_allowed = list(getattr(policy, "stacks_allowed", []))
        if not repos_allowed:
            recommendations.append("Policy con repos_allowed vacio (DENY ALL): no se ejecutaran tareas de repo hasta definir allowlist.")
        if not stacks_allowed:
            recommendations.append("Policy con stacks_allowed vacio (DENY ALL deploy stacks).")

    if logs_summary.get("log_files_count", 0) == 0:
        recommendations.append("No hay logs de tareas aun: ejecuta un task de smoke test para poblar auditoria.")
    elif logs_summary.get("total_log_size_bytes", 0) > 30 * 1024 * 1024:
        recommendations.append("Logs recientes >30MB: planificar rotacion/retencion de logs.")

    return recommendations


def _markdown_report(report: dict[str, Any]) -> str:
    metrics = report.get("metrics", {})
    status = report.get("status", {})
    recs = report.get("recommendations", [])
    lines = [
        "# AutoTuneAgent Report",
        "",
        f"- generated_at: {report.get('generated_at')}",
        f"- queue_depth: {metrics.get('queue_depth')}",
        f"- agents_alive: {status.get('agents_alive')}",
        f"- agents_missing: {status.get('agents_missing')}",
        f"- log_files_count: {metrics.get('log_files_count')}",
        "",
        "## Recommendations",
    ]
    for item in recs:
        lines.append(f"- {item}")
    lines.append("")
    return "\n".join(lines)


def _write_report(artifacts_dir: Path, report: dict[str, Any]) -> tuple[Path, Path]:
    out_dir = (artifacts_dir / "auto_improve").resolve()
    out_dir.mkdir(parents=True, exist_ok=True)

    ts_token = utc_now().strftime("%Y%m%dT%H%M%SZ")
    report_json_path = out_dir / f"report_{ts_token}.json"
    report_md_path = out_dir / f"report_{ts_token}.md"
    latest_json_path = out_dir / "latest_report.json"
    latest_md_path = out_dir / "latest_report.md"

    report_json_path.write_text(json.dumps(report, ensure_ascii=True, indent=2), encoding="utf-8")
    report_md_path.write_text(_markdown_report(report), encoding="utf-8")
    latest_json_path.write_text(json.dumps(report, ensure_ascii=True, indent=2), encoding="utf-8")
    latest_md_path.write_text(_markdown_report(report), encoding="utf-8")
    return latest_json_path, latest_md_path


def run_cycle(args: argparse.Namespace) -> int:
    config_dir = Path(args.config_dir).resolve()
    logs_dir = Path(args.logs_dir).resolve()
    artifacts_dir = Path(args.artifacts_dir).resolve()

    queue = RedisQueue(redis_url=args.redis_url, queue_name=args.queue_name)
    queue.ping()

    policy = None
    try:
        policy = load_policy(config_dir / "policy.yaml")
    except (PolicyError, OSError) as exc:
        print(f"[autotune][WARN] failed to load policy: {exc}")

    queue_depth = queue.length()
    status_snapshot = _load_status_snapshot(queue, str(args.status_redis_key))
    logs_summary = _summarize_recent_logs(logs_dir, max_files=args.max_log_files)
    recommendations = _build_recommendations(
        policy=policy,
        queue_depth=queue_depth,
        status_snapshot=status_snapshot,
        logs_summary=logs_summary,
    )

    agents = status_snapshot.get("agents")
    agents_alive = 0
    agents_missing = 0
    if isinstance(agents, list):
        for item in agents:
            if isinstance(item, dict):
                if bool(item.get("alive")):
                    agents_alive += 1
                else:
                    agents_missing += 1

    report = {
        "generated_at": utc_now_iso(),
        "agent": "AutoTuneAgent",
        "metrics": {
            "queue_depth": queue_depth,
            "log_files_count": logs_summary.get("log_files_count", 0),
            "total_log_size_bytes": logs_summary.get("total_log_size_bytes", 0),
            "newest_logs": logs_summary.get("newest_logs", []),
        },
        "status": {
            "agents_alive": agents_alive,
            "agents_missing": agents_missing,
            "maintainer_snapshot_ts": status_snapshot.get("ts"),
        },
        "recommendations": recommendations,
    }

    latest_json_path, latest_md_path = _write_report(artifacts_dir, report)
    queue.publish_heartbeat(
        agent="AutoTuneAgent",
        linux_user=os.environ.get("CLAWBRAIN_AUTOTUNE_LINUX_USER", "codex"),
        ttl_seconds=max(20, int(args.interval_seconds) * 2),
        extra={
            "state": "idle",
            "report_json": str(latest_json_path),
            "report_md": str(latest_md_path),
        },
    )
    print(
        f"[autotune] generated_at={report['generated_at']} queue_depth={queue_depth} "
        f"report={latest_json_path}"
    )
    return 0


def main() -> int:
    args = parse_args()
    interval = max(30, int(args.interval_seconds))

    while True:
        try:
            code = run_cycle(args)
        except QueueError as exc:
            print(f"[autotune][FAIL] redis error: {exc}")
            code = 1
        except OSError as exc:
            print(f"[autotune][FAIL] io error: {exc}")
            code = 1
        except Exception as exc:  # noqa: BLE001
            print(f"[autotune][FAIL] unexpected: {exc}")
            code = 1

        if args.once:
            return code

        time.sleep(interval)


if __name__ == "__main__":
    raise SystemExit(main())

File: src/runner/autotune.ts
import fs from "node:fs";
import path from "node:path";

import { Command } from "commander";

import { QueueError, PolicyError } from "../errors";
import { loadPolicy } from "../policy";
import { RedisQueue } from "../queue";

const DEFAULT_CONFIG_DIR = "/data/clawbrain/config";
const DEFAULT_LOGS_DIR = "/data/clawbrain/logs";
const DEFAULT_ARTIFACTS_DIR = "/data/clawbrain/artifacts";
const DEFAULT_REDIS_URL = "redis://127.0.0.1:6379/0";
const DEFAULT_QUEUE_NAME = "clawbrain:tasks";
const DEFAULT_STATUS_REDIS_KEY = "clawbrain:agent_status:latest";

interface AutoTuneArgs {
  configDir: string;
  logsDir: string;
  artifactsDir: string;
  redisUrl: string;
  queueName: string;
  statusRedisKey: string;
  intervalSeconds: number;
  maxLogFiles: number;
  once: boolean;
}

function parseArgs(argv: string[]): AutoTuneArgs {
  const program = new Command();
  program
    .option("--config-dir <dir>", "Config dir", process.env.CLAWBRAIN_CONFIG_DIR ?? DEFAULT_CONFIG_DIR)
    .option("--logs-dir <dir>", "Logs dir", process.env.CLAWBRAIN_LOGS_DIR ?? DEFAULT_LOGS_DIR)
    .option("--artifacts-dir <dir>", "Artifacts dir", process.env.CLAWBRAIN_ARTIFACTS_DIR ?? DEFAULT_ARTIFACTS_DIR)
    .option("--redis-url <url>", "Redis URL", process.env.CLAWBRAIN_REDIS_URL ?? DEFAULT_REDIS_URL)
    .option("--queue-name <name>", "Queue name", process.env.CLAWBRAIN_QUEUE_NAME ?? DEFAULT_QUEUE_NAME)
    .option(
      "--status-redis-key <key>",
      "Maintainer status key",
      process.env.CLAWBRAIN_AGENT_STATUS_REDIS_KEY ?? DEFAULT_STATUS_REDIS_KEY,
    )
    .option(
      "--interval-seconds <sec>",
      "Run interval",
      process.env.CLAWBRAIN_AUTOTUNE_INTERVAL_SEC ?? "180",
    )
    .option(
      "--max-log-files <n>",
      "Max recent log files",
      process.env.CLAWBRAIN_AUTOTUNE_MAX_LOG_FILES ?? "40",
    )
    .option("--once", "Run one cycle", false);

  program.parse(argv);
  const opts = program.opts<Record<string, string | boolean>>();
  return {
    configDir: String(opts.configDir),
    logsDir: String(opts.logsDir),
    artifactsDir: String(opts.artifactsDir),
    redisUrl: String(opts.redisUrl),
    queueName: String(opts.queueName),
    statusRedisKey: String(opts.statusRedisKey),
    intervalSeconds: Number(opts.intervalSeconds),
    maxLogFiles: Number(opts.maxLogFiles),
    once: Boolean(opts.once),
  };
}

function summarizeRecentLogs(logsDir: string, maxFiles: number): Record<string, unknown> {
  const root = path.resolve(logsDir);
  if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) {
    return {
      log_files_count: 0,
      total_log_size_bytes: 0,
      newest_logs: [],
    };
  }

  const entries = fs
    .readdirSync(root)
    .filter((name) => name.endsWith(".log"))
    .map((name) => {
      const absolute = path.join(root, name);
      const stat = fs.statSync(absolute);
      return { name, absolute, stat };
    })
    .filter((entry) => entry.stat.isFile())
    .sort((a, b) => b.stat.mtimeMs - a.stat.mtimeMs);

  const newest = entries.slice(0, Math.max(1, Math.floor(maxFiles)));
  let totalSize = 0;
  const newestLogs = newest.slice(0, 10).map((entry) => {
    totalSize += entry.stat.size;
    return {
      name: entry.name,
      size_bytes: entry.stat.size,
      mtime: new Date(entry.stat.mtimeMs).toISOString(),
    };
  });

  return {
    log_files_count: entries.length,
    total_log_size_bytes: totalSize,
    newest_logs: newestLogs,
  };
}

async function loadStatusSnapshot(queue: RedisQueue, statusRedisKey: string): Promise<Record<string, unknown>> {
  try {
    const raw = await queue.getRaw(statusRedisKey);
    if (!raw) {
      return {};
    }
    const parsed = JSON.parse(raw);
    if (!parsed || typeof parsed !== "object" || Array.isArray(parsed)) {
      return {};
    }
    return parsed as Record<string, unknown>;
  } catch {
    return {};
  }
}

function containsAny(text: string, keywords: string[]): boolean {
  const low = text.toLowerCase();
  return keywords.some((keyword) => low.includes(keyword));
}

function buildRecommendations(params: {
  policy: ReturnType<typeof loadPolicy> | null;
  queueDepth: number;
  statusSnapshot: Record<string, unknown>;
  logsSummary: Record<string, unknown>;
}): string[] {
  const recommendations: string[] = [];
  const missingAgents = Array.isArray(params.statusSnapshot.missing_agents)
    ? (params.statusSnapshot.missing_agents as unknown[]).map((value) => String(value))
    : [];

  if (params.queueDepth >= 5) {
    recommendations.push(
      "Queue depth >= 5: activar profile reinforce para escalar runner-coder-r2 y runner-builder-r2.",
    );
  } else if (params.queueDepth > 0) {
    recommendations.push("Hay tareas en cola: monitorear tiempos y mantener AutoTuneAgent activo.");
  } else {
    recommendations.push("Cola en cero: estado estable.");
  }

  if (missingAgents.length > 0) {
    recommendations.push(
      `Agentes faltantes reportados por maintainer: ${missingAgents.join(", ")}.`,
    );
  } else {
    recommendations.push("Maintainer reporta agentes requeridos en linea.");
  }

  if (!params.policy) {
    recommendations.push("No se pudo cargar policy activa, validar /data/clawbrain/config/policy.yaml.");
  } else {
    if ((params.policy.repos_allowed ?? []).length === 0) {
      recommendations.push(
        "Policy con repos_allowed vacio (DENY ALL): no se ejecutaran tareas de repo hasta definir allowlist.",
      );
    }
    if ((params.policy.stacks_allowed ?? []).length === 0) {
      recommendations.push("Policy con stacks_allowed vacio (DENY ALL deploy stacks).");
    }
  }

  const logFilesCount = Number(params.logsSummary.log_files_count ?? 0);
  const totalLogSize = Number(params.logsSummary.total_log_size_bytes ?? 0);
  if (logFilesCount === 0) {
    recommendations.push("No hay logs de tareas aun: ejecuta un task de smoke test para poblar auditoria.");
  } else if (totalLogSize > 30 * 1024 * 1024) {
    recommendations.push("Logs recientes >30MB: planificar rotacion/retencion de logs.");
  }

  return recommendations;
}

function markdownReport(report: Record<string, unknown>): string {
  const metrics = (report.metrics as Record<string, unknown>) ?? {};
  const status = (report.status as Record<string, unknown>) ?? {};
  const recs = Array.isArray(report.recommendations)
    ? (report.recommendations as unknown[]).map((item) => String(item))
    : [];

  const lines: string[] = [
    "# AutoTuneAgent Report",
    "",
    `- generated_at: ${String(report.generated_at ?? "")}`,
    `- queue_depth: ${String(metrics.queue_depth ?? "")}`,
    `- agents_alive: ${String(status.agents_alive ?? "")}`,
    `- agents_missing: ${String(status.agents_missing ?? "")}`,
    `- log_files_count: ${String(metrics.log_files_count ?? "")}`,
    "",
    "## Recommendations",
  ];

  for (const item of recs) {
    lines.push(`- ${item}`);
  }
  lines.push("");
  return lines.join("\n");
}

function writeReport(artifactsDir: string, report: Record<string, unknown>): { latestJson: string; latestMd: string } {
  const outDir = path.resolve(path.join(artifactsDir, "auto_improve"));
  fs.mkdirSync(outDir, { recursive: true });

  const now = new Date();
  const tsToken = now.toISOString().replace(/[-:]/g, "").replace(/\.\d{3}Z$/, "Z");

  const reportJson = path.join(outDir, `report_${tsToken}.json`);
  const reportMd = path.join(outDir, `report_${tsToken}.md`);
  const latestJson = path.join(outDir, "latest_report.json");
  const latestMd = path.join(outDir, "latest_report.md");

  const reportJsonText = `${JSON.stringify(report, null, 2)}\n`;
  const reportMdText = `${markdownReport(report)}\n`;

  fs.writeFileSync(reportJson, reportJsonText, { encoding: "utf-8" });
  fs.writeFileSync(reportMd, reportMdText, { encoding: "utf-8" });
  fs.writeFileSync(latestJson, reportJsonText, { encoding: "utf-8" });
  fs.writeFileSync(latestMd, reportMdText, { encoding: "utf-8" });

  return { latestJson, latestMd };
}

async function runCycle(args: AutoTuneArgs): Promise<number> {
  const queue = new RedisQueue({ redisUrl: args.redisUrl, queueName: args.queueName });
  await queue.ping();

  let policy: ReturnType<typeof loadPolicy> | null = null;
  try {
    policy = loadPolicy(path.join(path.resolve(args.configDir), "policy.yaml"));
  } catch (error) {
    if (!(error instanceof PolicyError)) {
      // eslint-disable-next-line no-console
      console.warn(`[autotune][WARN] failed to load policy: ${String(error)}`);
    }
  }

  const queueDepth = await queue.length();
  const statusSnapshot = await loadStatusSnapshot(queue, args.statusRedisKey);
  const logsSummary = summarizeRecentLogs(args.logsDir, args.maxLogFiles);
  const recommendations = buildRecommendations({
    policy,
    queueDepth,
    statusSnapshot,
    logsSummary,
  });

  const agents = Array.isArray(statusSnapshot.agents)
    ? (statusSnapshot.agents as unknown[]).filter(
        (value) => value && typeof value === "object" && !Array.isArray(value),
      )
    : [];
  let agentsAlive = 0;
  let agentsMissing = 0;
  for (const item of agents as Array<Record<string, unknown>>) {
    if (Boolean(item.alive)) {
      agentsAlive += 1;
    } else {
      agentsMissing += 1;
    }
  }

  const report: Record<string, unknown> = {
    generated_at: new Date().toISOString(),
    agent: "AutoTuneAgent",
    metrics: {
      queue_depth: queueDepth,
      log_files_count: Number(logsSummary.log_files_count ?? 0),
      total_log_size_bytes: Number(logsSummary.total_log_size_bytes ?? 0),
      newest_logs: logsSummary.newest_logs ?? [],
    },
    status: {
      agents_alive: agentsAlive,
      agents_missing: agentsMissing,
      maintainer_snapshot_ts: statusSnapshot.ts,
    },
    recommendations,
  };

  const reportPaths = writeReport(args.artifactsDir, report);
  await queue.publishHeartbeat({
    agent: "AutoTuneAgent",
    linuxUser: process.env.CLAWBRAIN_AUTOTUNE_LINUX_USER ?? "codex",
    ttlSeconds: Math.max(20, Math.floor(args.intervalSeconds) * 2),
    extra: {
      state: "idle",
      report_json: reportPaths.latestJson,
      report_md: reportPaths.latestMd,
    },
  });

  // eslint-disable-next-line no-console
  console.log(
    `[autotune] generated_at=${String(report.generated_at)} queue_depth=${queueDepth} report=${reportPaths.latestJson}`,
  );

  return 0;
}

async function main(argv: string[]): Promise<number> {
  const args = parseArgs(argv);
  const interval = Math.max(30, Math.floor(args.intervalSeconds));

  while (true) {
    let code = 0;
    try {
      code = await runCycle(args);
    } catch (error) {
      if (error instanceof QueueError) {
        // eslint-disable-next-line no-console
        console.error(`[autotune][FAIL] redis error: ${error.message}`);
      } else {
        // eslint-disable-next-line no-console
        console.error(`[autotune][FAIL] unexpected: ${String(error)}`);
      }
      code = 1;
    }

    if (args.once) {
      return code;
    }

    await new Promise((resolve) => setTimeout(resolve, interval * 1000));
  }
}

if (require.main === module) {
  void main(process.argv).then((code) => process.exit(code));
}
